hypothesis-python/src/hypothesis/internal/conjecture/dfa/__init__.py

```diff
@@ -116,61 +119,199 @@ def all_matching_regions(self, string):
+                if k in stack_set:
+                    # We should never have put a dead node on the
+                    # stack in the first place.
+                    assert not self.is_dead(k)
+                    cache[k] = inf
+                    break
+                elif k not in cache and not self.is_dead(k):
+                    stack.append(k)
+                    stack_set.add(k)
+                    break
+            else:
+                cache[j] = max(
+                    (
+                        1 + cache[k]
+                        for k in self.successor_states(j)
+                        if not self.is_dead(k)
+                    ),
+                    default=0,
+                )
+
+                # j is live so it must either be accepting or have a live child.
+                assert self.is_accepting(j) or cache[j] > 0
+                pop()
+        return cache[i]
+
+    def count_strings(self, state, length):
         """Returns the number of strings of length ``k``
         that are accepted when starting from state ``i``."""
-        assert k >= 0
-        if k == 0:
-            if self.is_accepting(i):
-                return 1
+        assert length >= 0
+        cache = self.__cache("count_strings")
+
+        try:
+            return cache[state, length]
+        except KeyError:
+            pass
+
+        pending = [(state, length)]
+        seen = set()
+        i = 0
+
+        while i < len(pending):
+            s, n = pending[i]
+            i += 1
+            if n > 0:
+                for t in self.successor_states(s):
+                    key = (t, n - 1)
+                    if key not in cache and key not in seen:
+                        assert key not in pending
+                        pending.append(key)
+                        seen.add(key)
+
+        while pending:
+            s, n = pending.pop()
+            if n == 0:
+                cache[s, n] = int(self.is_accepting(s))
             else:
-                return 0
-        if k > self.max_length(i):
-            return 0
-        return sum(self.count_strings(j, k - 1) for _, j in self.transitions(i))
+                cache[s, n] = sum(cache[t, n - 1] for _, t in self.transitions(s))
 
-    @cached
-    def reachable(self, i):
-        """Returns the set of all states reachable
-        by traversing some non-empty string starting from
-        state i."""
-        reached = set()
+        return cache[state, length]
 
-        queue = deque([i])
+    @cached
+    def successor_states(self, state):
+        """Returns all of the distinct states that can be reached via one
+        transition from ``state``, in the lexicographic order of the
+        smallest character that reaches them."""
+        seen = set()
+        result = []
+        for _, j in self.raw_transitions(state):
+            if j not in seen:
+                seen.add(j)
+                result.append(j)
+        return tuple(result)
```

Comment:
It's mildly annoying that `ConcreteDFA.raw_transitions` goes out of its way to decompress its transition table, and then we turn around and throw most of that work away here.

(I'm fine with keeping things simple and doing nothing for now; just felt it was worth mentioning.)

Reply:
I agree! I'm favour keeping things simple and doing nothing for now, but it would be nice to improve the logic there.